[4~9단계 - 방탈출 예약 관리] 헤일러(민서현) 미션 제출합니다.

threepebbles opened the pull request:
안녕하세요 피케이!

우선 LMS에서 제시해준 학습 키워드를 위주로 Spring 학습을 진행했고, 

패키지 구조, 각 계층간 의존 관계, 예외 처리 위치와 방식 등 레벨1에서 연습했던 내용들을 같이 적용해보려고 노력했습니다.

리뷰 잘 부탁드립니다! 😃

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [헤일러의 프롤로그](https://prolog.techcourse.co.kr/studylogs/4217)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
✅: 잘 지킴, ❓: 애매함
<!-- 이유 작성 -->
규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다. ✅
규칙 2: else 예약어를 쓰지 않는다. ✅
규칙 3: 모든 원시값과 문자열을 포장한다. ❓
규칙 4: 한 줄에 점을 하나만 찍는다. ✅
규칙 5: 줄여쓰지 않는다(축약 금지). ✅
규칙 6: 모든 엔티티를 작게 유지한다. ✅
규칙 7: 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다. ❓
규칙 8: 일급 콜렉션을 쓴다. ✅
규칙 9: 게터/세터/프로퍼티를 쓰지 않는다. ❓

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [x] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?

## 📌 코드 스타일 컨벤션

https://github.com/woowacourse/spring-roomescape-admin/pull/260#discussion_r2049992494

> convention 을 git 으로 관리해주시네요 😮
> 
> 
> 이렇게 하게 된 계기가 있나요?
> 

이전에 컨벤션을 git으로 관리하는 것에 대해 커멘트를 남겨 주셨었는데요. 

git pull을 하면서 IDE에 자동으로 컨벤션이 반영되도록 하는 것이 아니라면 항상 싱크에 문제가 생길 수 있는 가능성이 존재하기 때문에 지금의 공유 방식은 적절하지 않을 수 있다는 생각이 들었습니다. 그래서

1. git pull 시 자동 반영이 되도록 하는 도구를 찾아보거나, 
2. 다른 협업 문서 관리 도구로 깃 브랜치 전략, 커밋 컨벤션과 같은 내용들과 함께 컨벤션을 관리하는 것이 더 괜찮은 방식이 아닐까 고민을 해보았는데요, 

혹시 피케이는 어떤 접근이 현실적이라고 생각하시나요?

## 📌 `@Qualifier`의 사용

```java
@Service
public class ReservationService {

    private final ReservationRepository reservationRepository;
    private final ReservationTimeService reservationTimeService;

    public ReservationService(
            @Qualifier("h2ReservationRepository") final ReservationRepository reservationRepository,
            final ReservationTimeService reservationTimeService
    ) {
        this.reservationRepository = reservationRepository;
        this.reservationTimeService = reservationTimeService;
    }
    ...
```

ReservationRepository 인터페이스와 여러 개의 Repository 구현체를 두고, 구현체들을 교체할 수 있게 하는 방법으로 `@Qualifier`를 사용했습니다.

`@Qualifier` 를 사용하게 되면 어떤 Repository 구현체를 주입할 지 Service 계층에서 바로 파악이 가능하고, 다른 구현체로 교체하고 싶을 때 코드의 수정이 직관적이라는 장점이 있어 사용했습니다. 그리고 IntelliJ IDE에서 Qualifier 내의 Bean 이름이 등록된 Bean 이름인지 검사도 해주어서 오타에 대한 걱정도 없어서 사용하기 편했던 것 같습니다.

`@Qualifier` 대신 `@Primary` 를 사용하는 방법도 있었는데요! `@Primary`를 사용하면 코드는 더 간결하지만, 어느 Repository 구현체에 `@Primary` 어노테이션이 붙어있는지 파일을 일일히 확인해야 하는 번거로움이 더 큰 단점으로 다가와서 사용하지 않았습니다.

`@Primary` 보다 `@Qualifier`를 선택한 것이 적절한 선택이었는지, 유지보수의 관점이 아닌 다른 관점(ex. Service에서 구체 클래스의 종류를 알게 되는 것)에서 문제가 발생할 수 있을지 궁금합니다!

## 📌 queryForObject() 사용을 지양하고, query() 사용

```java
@Override
public Optional<ReservationTime> findById(final Long id) {
    String sql = "select * from reservation_times where id = ?";
    List<ReservationTime> reservationTimes = jdbcTemplate.query(sql, (rs, rowNum) ->
                    new ReservationTime(rs.getLong("id"),
                            rs.getTime("start_at").toLocalTime()),
            id
    );

    if (reservationTimes.size() == 1) {
        return Optional.of(reservationTimes.getFirst());
    }
    return Optional.empty();
}
```

제가 생각하기에 데이터 단건 조회 시 데이터가 없는 것 또한 정상적인 조회의 결과일 수 있으므로 예외를 터뜨리는 것이 적절하지 않다고 생각이 들었습니다. 

`queryForObject()` 를 사용하면 조회 쿼리의 결과가 1건이 아니면 예외가 발생(결과가 0개 -> `EmptyResultDataAccessException`, 결과가 2개 이상 -> `IncorrectResultSizeDataAccessException`)하므로 반드시 예외 처리를 해서 Optional.empty()를 반환하도록 해야 했습니다. 

그래서 예외를 발생시키고 catch하는 방식보다는 처음부터 예외가 아예 발생하지 않는 `query()` 를 사용해서 `size()` 검사를 하는 것이 더 안전하고, 가독성 좋고(예외 처리 로직이 사라졌으므로), 성능 측면에서도 유리하다고 생각이 들었습니다.

피케이는 어떻게 생각하시나요?

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

threepebbles committed:
chore(gradle): jdbc 의존성 추가

threepebbles committed:
chore: h2 데이버테이스 console 기능 활성화, datasource url 지정

threepebbles committed:
chore: application.properties대신 application.yml 사용, DB 스키마 추가

threepebbles committed:
test: DB명 검증, 예약 테이블명 검증 테스트 추가

threepebbles committed:
test: 예약 API 관련 테스트 메서드를 UserReservationRestControllerTest로 이동

threepebbles committed:
chore: 테스트 DB 스키마 추가

threepebbles committed:
chore: 테스트 스프링 설정 추가

threepebbles committed:
feat(Reservation): 생성자 추가, JsonCreater로 직렬화에 사용할 생성자 명확하게 지정

threepebbles committed:
feat: h2 데이터베이스를 이용한 예약 정보 저장 관리 기능 추가

threepebbles committed:
chore: 프로덕션 DB 스키마와 테스트 DB 스키마 동일하게 수정

threepebbles committed:
refactor: 클래스명 변경

threepebbles committed:
chore: 시간 테이블 추가

threepebbles committed:
refactor: RequestMapping 사용해서 중복 코드 제거

threepebbles committed:
chore: 예약 시간 테이블명 변경

threepebbles committed:
refactor: 패키지 이동

threepebbles committed:
refactor: 변수명 변경

threepebbles committed:
chore: 예약 시간 테이블이 존재하지 않는 경우만 생성하도록 수정

threepebbles committed:
docs: 예약 시간 기능 목록 추가

threepebbles committed:
feat: 예약 시간 관리 기능(추가, 조회, 제거) 추가

threepebbles committed:
chore: 예약 페이지 파일 수정

threepebbles committed:
feat: 예약 추가 요청 dto, 예약 조회 응답 dto 추가

threepebbles committed:
chore: reservation 테이블의 time 컬럼과 reservation_times 테이블의 start_at 컬럼 외래 키 관계 설정

- 예약 시간 중복 방지 목적

threepebbles committed:
fix: 테이블 생성 순서 수정

threepebbles committed:
feat: 시간으로 예약 시간 데이터 조회하는 기능 추가

threepebbles committed:
feat(ReservationTimeService): 예약 시간 관리 서비스 계층 추가

threepebbles committed:
remove: 사용하지 않는 기능 제거

- 요청 dto를 도메인 객체로 변환하는 기능 제거

threepebbles committed:
feat(ReservationService): 예약 정보 관리 서비스 계층 추가

threepebbles committed:
remove: dto 계층 분리로 인해 더 이상 사용하지 않는 어노테이션 제거

threepebbles committed:
refactor: 생성자에서 @Qualifier를 이용해 주입할 빈을 명시하도록 수정

threepebbles committed:
test: 예약 시간을 추가하는 API 테스트 추가

threepebbles committed:
refactor: 테스트 코드 가독성 개선

threepebbles committed:
refactor: 중복 코드 제거

threepebbles committed:
feat: 예약 추가 API 명세 수정

- LocalTime time -> Long time_id

threepebbles committed:
refactor: 데이터 조회 시, 데이터가 존재하지 않는 예외를 발생시키는 책임을 Repository 계층에서 Service 계층으로 위임

threepebbles committed:
test: Reservation 생성자 검증 테스트 추가

threepebbles committed:
test: ReservationTime 생성자 검증 테스트 추가

threepebbles committed:
test: 컨트롤러에 JdbcTemplate 타입의 필드가 존재하는지 검증하는 테스트 추가

threepebbles committed:
feat(GlobalExceptionHandler): ControllerAdvice를 이용해 Controller의 공통 예외를 처리하도록 수정

pkeugine commented on src/main/java/roomescape/admin/repository/time/H2ReservationTimeRepository.java:
```diff
@@ -0,0 +1,63 @@
+                id
+        );
+
+        if (reservationTimes.size() == 1) {
```
> 제가 생각하기에 데이터 단건 조회 시 데이터가 없는 것 또한 정상적인 조회의 결과일 수 있으므로 예외를 터뜨리는 것이 적절하지 않다고 생각이 들었습니다.
> 
> queryForObject() 를 사용하면 조회 쿼리의 결과가 1건이 아니면 예외가 발생(결과가 0개 -> EmptyResultDataAccessException, 결과가 2개 이상 -> IncorrectResultSizeDataAccessException)하므로 반드시 예외 처리를 해서 Optional.empty()를 반환하도록 해야 했습니다.
> 
> 그래서 예외를 발생시키고 catch하는 방식보다는 처음부터 예외가 아예 발생하지 않는 query() 를 사용해서 size() 검사를 하는 것이 더 안전하고, 가독성 좋고(예외 처리 로직이 사라졌으므로), 성능 측면에서도 유리하다고 생각이 들었습니다.
>
> 피케이는 어떻게 생각하시나요?

`query()` 를 사용해서 `size()`검사하는 것 좋네요 👍

`id`가 primary key라서 그럴 경우가 없을 것 같긴 하지만,
결과가 2개 이상 나왔을 경우 데이터를 놓치게 됩니다. 
"값이 있음"을 검사하는 조건문으로 변경하는 것이 더 명확할 것 같아요.

threepebbles replied:
```suggestion
        if (!reservationTimes.isEmpty()) {
```
size()==1보다는 empty()를 확인하는 방식으로 검증하면 좀 더 명확해질까요?

threepebbles replied:
저는 id가 unique함을 가정하고 조회가 0건, 1건인 경우만 고려했었는데요!
2개 이상의 결과가 조회된 경우는 트랜잭션의 일관성에 어긋나는 예측 밖의 결과라서 예외가 발생해야 하지 않을까 싶습니다...!

pkeugine replied:
기존의 구현에서 결과가 2개 이상 나왔을 때 데이터를 놓치게 된다는 점만 유의해서 헤일러가 판단 하시기에 더 명확한 방법으로 구현 해주시면 됩니다. :)

pkeugine commented on src/main/java/roomescape/admin/repository/time/H2ReservationTimeRepository.java:
```diff
@@ -0,0 +1,63 @@
+            return ps;
+        }, keyHolder);
+
+        return keyHolder.getKey().longValue();
```
`keyHolder.getKey()` 가 `null` 일경우 `NPE`가 발생합니다.

`insert`는 되었지만 `key`를 반환하지 못한 경우,
실제로 `insert` 한 `row`가 없는 경우 등이 있을 수 있어서 `keyHolder.getKey()`이 `null`일 가능성도 배제할 순 없습니다. 
이 부분도 같이 고려해보죠.

threepebbles replied:
insert가 성공하더라도 
1. prepareStatement()에서 키 반환 설정이 잘못되었거나
```java
connection.prepareStatement(sql, new String[]{"오타"});
```
2. DB에서 자동 키 생성 기능을 제공을 안 하는 경우 

`keyholder.getKey()`가 null일 가능성이 있을 수 있겠네요!

null인 경우에 예외를 발생시키도록 수정했습니다 :)

pkeugine commented on src/main/java/roomescape/admin/service/ReservationTimeService.java:
```diff
@@ -0,0 +1,45 @@
+    private final ReservationTimeRepository reservationTimeRepository;
+
+    public ReservationTimeService(
+            @Qualifier("h2ReservationTimeRepository") final ReservationTimeRepository reservationTimeRepository) {
```
**Q. `@Primary` 보다 `@Qualifier`를 선택한 것이 적절한 선택이었는지, 유지보수의 관점이 아닌 다른 관점(ex. Service에서 구체 클래스의 종류를 알게 되는 것)에서 문제가 발생할 수 있을지**

개발을 하다보면 늘 100% 정답은 없는 것 같습니다.
장점과 단점이 있다면 장점을 최대한 살릴 수 있는 방식으로 만들어나가는 듯 하네요.
헤일러가 남겨주신 내용에 공감합니다. 특정 구현체 의존성을 높이지만, 유지보수라는 실질적인 이점을 챙기는거죠.
그런 면에서 `@Primary`보다 `@Qualifier`를 선택한 것은 적절했다고 생각합니다.

그렇다면 아래 처럼 직접 `H2ReservationRepository`를 가져오는 방법과 비교해보면 어떨까요?
 `@Qualifier` 를 사용하는 것과 아래 처럼 구현하는 것은
`Service`가 구체 클래스의 종류를 알게 된다는 관점에서 동일하죠 :)
이 방식이 정답이라는 것은 아니고, 생각해보면 재밌을 것 같았습니다.

```java
public class ReservationService {

    private final H2ReservationRepository reservationRepository;
    private final ReservationTimeService reservationTimeService;

    public ReservationService(
            final H2ReservationRepository reservationRepository,
            final ReservationTimeService reservationTimeService
    ) {
        this.reservationRepository = reservationRepository;
        this.reservationTimeService = reservationTimeService;
    }
}
```

더 나아가, 현재 `InMemoryReservationRepository`를 사용하는 곳이 없는데
계속 남겨둠으로써 구현이 복잡해진 것 같습니다. 🤔
 `InMemoryReservationRepository`가 더 이상 필요하지 않으면 제거해도 무방하다고 생각합니다.
필요할 때 다시 추가하면 됩니다. git 으로 히스토리까지 다 관리되고 있고요.

threepebbles replied:
> InMemoryReservationRepository가 더 이상 필요하지 않으면 제거해도 무방하다고 생각합니다.

`InMemoryReservationRepository`는 사용하는 곳이 없으므로 제거했습니다!

> 그렇다면 아래 처럼 직접 H2ReservationRepository를 가져오는 방법과 비교해보면 어떨까요?

둘의 가장 큰 차이는 Qualifier를 사용하는 경우, 구체 클래스 자체에 의존하는 것이 아닌 `등록한 Bean 이름에만 의존하는 것`이라고 생각합니다.

`@Qualifier("h2ReservationRepository")`라는 코드가 `H2ReservationRepository` 클래스에 직접 의존하는 것처럼 보이지만 사실은 `h2ReservationRepository`라는 Bean 이름에 의존하고 있는 것입니다.

제 생각에는 현재 코드에서 `@Qualifier("reservationRepository")`와 같이 수정하고, `reservationRepository`라는 이름의 Bean에 대응되는 클래스를 자유롭게 수정할 수 있는 설정 파일을 만든다면, 추후 `ReservationRepository` 구현 클래스를 교체해야 하는 상황이 왔을 때 `ReservationService` 코드를 하나도 수정하지 않고 `ReservationRepository` 구현체 클래스를 바꿀 수 있는 장점을 가져갈 수 있을 것 같습니다 😄

pkeugine replied:
구체적인 답변 좋네요. 맞습니다.
다만 `InMemoryReservationRepository` 를 제거했으니 `ReservationRepository` 의 구현체가 한 개 입니다.
따라서 `@Qualifier`를 사용하지 않아도 스프링이 알아서 `H2ReservationRepository` 로 주입해줍니다.
더 이상 필요하지 않은 `@Qualifier는` 제거해주죠 :)

threepebbles replied:
꼼꼼하게 봐주셔서 감사합니다 😄 
불필요한 Qualifier 어노테이션 제거했습니다!

pkeugine commented on src/main/java/roomescape/user/service/ReservationService.java:
```diff
@@ -0,0 +1,50 @@
+        if (found.isEmpty()) {
+            throw new DataNotFoundException("해당 예약 데이터가 존재하지 않습니다. id = " + id);
+        }
+        reservationRepository.delete(found.get());
```
`orElseThrow` 를 사용하면 아래와 같이 개선할 수 있습니다.

```suggestion
        Reservation found = reservationRepository.findById(id)
                .orElseThrow(() -> new DataNotFoundException("해당 예약 데이터가 존재하지 않습니다. id = " + id));

        reservationRepository.delete(found);
```

pkeugine commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,15 @@
+  sql:
+    init:
+      mode: embedded
+      schema-locations: classpath:schema.sql
```
파일마다 새로운 line 을 넣어주시기 바랍니다.

<img width="78" alt="스크린샷 2025-02-14 오전 2 34 06" src="https://github.com/user-attachments/assets/90b17a4b-ef1f-46b7-a4df-b36cf4c2328f" />

threepebbles replied:
POSIX 표준을 지키기 위해 파일 끝에 개행을 해야 하는군요!

![image](https://github.com/user-attachments/assets/64b6eaf6-5dd9-42fe-8ef9-bbf03636d39b)

IntelliJ 설정을 통해 자동으로 적용되도록 했습니다 😄

pkeugine replied:
좋습니다.
다만 아직 해당 표시가 남아있네요. 다시 확인해주세요.

threepebbles replied:
제가 눈으로 확인한 몇몇 파일만 수정해서 몇 개의 파일에서 코드 스타일 적용이 누락됐었네요..!
ctrl + alt + shift + L 기능을 통해 프로젝트 파일 전체에 적용했습니다!

[4253fa7](https://github.com/woowacourse/spring-roomescape-admin/pull/321/commits/4253fa72ebc5fdb89c7e866d404767f77ba07723)

pkeugine commented on src/main/java/roomescape/user/repository/reservation/H2ReservationRepository.java:
```diff
@@ -0,0 +1,102 @@
+                SELECT 
+                    r.id as id, 
+                    r.name as name, 
+                    date, 
```
`findById`와 `findAll` 내부 `sql`에서 `alias` 지정하는 것을 통일해주면 좋을 것 같습니다.

pkeugine commented on src/main/java/roomescape/user/repository/reservation/H2ReservationRepository.java:
```diff
@@ -0,0 +1,102 @@
+                        rs.getDate("date").toLocalDate(),
+                        new ReservationTime(
+                                rs.getLong("time_id"),
+                                LocalTime.parse(rs.getString("start_at"))
```
아래와 같이 변경할 수 있습니다.
```suggestion
                                rs.getTime("start_at").toLocalTime()
```

pkeugine commented on src/main/java/roomescape/user/repository/reservation/H2ReservationRepository.java:
```diff
@@ -0,0 +1,102 @@
+                WHERE r.id = ?
+                """;
+        List<Reservation> reservations = jdbcTemplate.query(sql, (rs, rowNum) ->
+                        new Reservation(
```
결과를 `Reservation` 객체로 매핑하는 코드가 중복되니 `RowMapper`를 추출해서 재사용하면 어떨까요?

pkeugine commented on src/test/java/roomescape/user/controller/ReservationRestControllerTest.java:
```diff
@@ -0,0 +1,155 @@
+        final Map<String, String> params = new HashMap<>();
+        params.put("name", "헤일러");
+        params.put("date", "2025-04-15");
+        params.put("timeId", "1");
```
`params` 를 만드는 코드가 중복되니 메소드로 추출하면 어떨까요?

pkeugine commented on src/test/java/roomescape/user/domain/ReservationTest.java:
```diff
@@ -0,0 +1,49 @@
+    @Test
+    void 날짜가_null이면_예외가_발생한다() {
+        // given
+        final String name = "피케이";
```
🙋‍♂️

pkeugine commented on src/test/java/roomescape/user/controller/ReservationRestControllerTest.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    @Test
+    void 예약_정보_삭제에_성공한_경우_ok를_반환한다() {
```
아래 `예약_정보를_삭제한다()` 와 거의 유사해서 하나의 테스트로 합쳐줘도 괜찮을 것 같습니다.

pkeugine commented on src/test/java/roomescape/user/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,101 @@
+        final LocalDate date = LocalDate.parse("2023-08-05");
+
+        // when & then
+        Assertions.assertThatCode(() -> reservationService.save(name, date, time_id))
```
`doesNotThrowAnyException()` 예외가 발생하지 않는 것을 확인하는 것도 의미 있지만, 응답값으로 검사하는 것이 더 명확할 것 같습니다.

threepebbles replied:
```java
        // when
        final Long savedId = reservationService.save(name, date, timeId);
        final Reservation found = reservationService.getById(savedId);

        // then
        SoftAssertions.assertSoftly(softly -> {
            softly.assertThat(savedId).isEqualTo(found.getId());
            softly.assertThat(name).isEqualTo(found.getName());
            softly.assertThat(date).isEqualTo(found.getDate());
            softly.assertThat(timeId).isEqualTo(found.getTime().getId());
            softly.assertThat(time).isEqualTo(found.getTime().getStartAt());
        });
```
위 코드와 같이 저장된 예약 정보를 불러와서 필드 값을 전부 비교하는 테스트를 진행하여 테스트의 신뢰도를 높이는 방법은 어떻게 생각하시나요?

threepebbles replied:
제 생각에는 `findById` 기능에 의존하고 테스트 비용이 조금 증가하는 대신, 원하는 데이터 값이 저장되었는지 더 확실하게 테스트할 수 있다는 장점이 있을 것 같습니다!

pkeugine commented on src/test/java/roomescape/user/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,101 @@
+    void 예약_정보를_저장한다() {
+        // given
+        final LocalTime time = LocalTime.parse("10:00");
+        final Long time_id = reservationTimeService.save(time);
```
`snake_case`가 있네요!

pkeugine commented on src/test/java/roomescape/user/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,101 @@
+
+        @Bean
+        public ReservationTimeService reservationTimeService(
+                @Qualifier("h2ReservationTimeRepository") ReservationTimeRepository reservationTimeRepository
```
`@TestConfiguration` 내부에서는 직접 `Bean`을 등록하는 것이라 `@Qualifier`가 불필요해 보입니다.

pkeugine commented on README.md:
**Q. 다른 협업 문서 관리 도구로 깃 브랜치 전략, 커밋 컨벤션과 같은 내용들과 함께 컨벤션을 관리하는 것**


재밌는 주제네요. 말씀해 주신 내용에 동의합니다.
한 팀이 여러 레포지토리를 담당하는 경우도 꽤 많아요.
프로젝트마다 설정을 맞출 수도 있겠지만,
컨벤션의 경우 잘 변하지 않고, 팀 단위로 결정되는 경우도 있어서 깃으로 관리하기엔 적절하지 않을 가능성이 높죠. 
(딱히 커밋 히스토리를 남길 내용도 아니고요.) 

문서로 남기거나, 설정 파일을 다운 받을 수 있게 하는게 오히려 관리하기 편할 듯 하네요. 
그리고 lint 관리 툴을 커밋 또는 머지 과정에서 실행해도 괜찮을 것 같고요. 
(우테코 미션에서는 도입할 필요는 없습니다.)

pkeugine commented on src/main/java/roomescape/admin/controller/AdminController.java:
```diff
@@ -15,6 +15,6 @@ public String home() {
     public String reservation() {
-        return "admin/reservation-legacy";
+        return "admin/reservation";
     }
```
[7단계 - 시간 관리 기능](https://techcourse.woowahan.com/s/y5znJihX/ls/T9SVCqWJ) 요구사항을 보면
`admin/time`접속 시 시관 관리 페이지가 나와야 하는데 이 부분이 빠져있습니다.

threepebbles replied:
실수로 기능 구현을 누락했었네요 🥲 
해당 기능 추가했습니다!

```java
@Controller
@RequestMapping("/admin")
public class AdminController {

    ...
    @GetMapping("time")
    public String time() {
        return "admin/time";
    }
}
```

pkeugine reviewed (changes requested):
안녕하세요 헤일리,
4~9 단계 구현하시느라 고생 하셨습니다!
코멘트 몇 가지와 질문에 대한 답변 남겼으니 확인 부탁드려요 :)

threepebbles committed:
docs: 예약 시간 관리 화면 명세 추가

threepebbles committed:
feat: 예약 시간 관리 화면 응답 기능 구현

threepebbles committed:
refactor: 패키지 구조 변경

threepebbles committed:
refactor: 조회 결과 개수를 size()가 아닌 isEmpty()로 검사하도록 수정

threepebbles committed:
style: SQL 문자열에서 사용하는 SQL 키워드 모두 대문자로 수정

threepebbles committed:
remove(InMemoryReservationRepository): 사용하지 않는 클래스 파일 삭제

threepebbles committed:
refactor: KeyHolder로 받아온 키값이 null인 경우 예외 처리 추가

threepebbles committed:
refactor: 예약 시간, 예약 정보 저장 시 영향받은 행의 수 검증 로직 추가

threepebbles commented:
에구 제가 PR 템플릿을 노션에 따로 적어놓고 작성하는데 1~3단계 커멘트에 대한 답변을 깜빡하고 안 올렸었네요..!

아래는 1~3단계에서 마지막에 주셨던 커멘트들에 대한 답변입니다!

# 1~3단계 커멘트에 대한 답변

▶️ https://github.com/woowacourse/spring-roomescape-admin/pull/260#discussion_r2051487984

> 예외 처리를 적용해보며 `@ExceptionHandler` 도입까지 👍
> 
> 
> 그런데 만약 다른 controller 에 있는 api 에도 이와 비슷한 예외 처리를 적용하고 싶다면 어떻게 해야할까요?
> 
> 한 번 고민해보죠.
> 

`@ControllerAdvice` 를 이용해서 Controller들에서 공통으로 발생할 수 있는 예외를 처리하도록 했습니다 😃

```java
@ControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<Void> handleIllegalArgumentException(final IllegalArgumentException e) {
        return ResponseEntity.badRequest().build();
    }

    @ExceptionHandler(DataNotFoundException.class)
    public ResponseEntity<Void> handleDataNotFoundException(final DataNotFoundException e) {
        return ResponseEntity.notFound().build();
    }
}
```

▶️ https://github.com/woowacourse/spring-roomescape-admin/pull/260#discussion_r2051488208

> `LocalDateTime` 이라는 클래스도 있는데, 이것도 한 번 알아보고 어떤 것을 사용하면 좋을지 고민해보아요.
> 

Reservation 클래스에서 LocalDate과 LocalTime을 묶어서 LocalDateTime으로 관리하게 된다면, 데이터베이스 테이블과 도메인 객체 사이의 괴리가 생길 것 같다는 생각이 듭니다. 

현재 제 코드에서는 도메인 모델과 데이터베이스 엔티티를 동일한 클래스로 사용했는데요!
둘을 다른 클래스로 분리한다면 도메인 모델에서 필드를 1개 줄일 수 있다는 이점이 있는 대신, 데이터베이스 엔티티↔도메인 모델 간 변환을 위한 Mapper 클래스가 추가되어 관리할 코드가 늘아난다는 단점이 있을 것 같습니다.

현재 상황에서 도메인 모델과 데이터베이스 엔티티를 분리하는 것이 큰 이점이 있을까요?

```java
// 데이터베이스 엔티티
public record ReservationEntity(Long id, String name, LocalDate date, ReservationTime time) {
}
```

```java
// 도메인 모델
public class Reservation {
    private final Long id;
    private final String name;
    private final LocalDateTime dateTime;
    ...
}
```

▶️ https://github.com/woowacourse/spring-roomescape-admin/pull/260#discussion_r2051488392

> @DirtiesContext 를 사용하지 않기로 한 이유가 궁금하네요.
어떤 방식으로 작동하길래 이 대신 @BeforeEach 를 사용하셨나요?
> 

`ApplicationContext` 는 Spring 애플리케이션의 모든 Bean의 설정 정보와 라이프 사이클을 관리하는 스프링의 인터페이스입니다.

`@SpringBootTest` 를 사용한 테스트 클래스 내에서는 같은 `ApplicationContext` 를 재활용합니다. (ContextCaching 기능) 그래서 `ApplicationContext` 에 등록된 Bean들의 상태가 변화하면 테스트 간 격리가 제대로 수행되지 않을 수 있습니다.

`@DirtiesContext` 를 사용하면 매 테스트마다 `ApplicationContext` 를 다시 생성하기 때문에 Bean의 상태 변화가 다른 테스트에 영향을 미치지 않게 만들 수 있습니다.

하지만 당시 Controller를 테스트하는데 Repository가 단순히 컬렉션에 데이터를 저장/조회하는 로직을 담당하고 있었고, 컬렉션을 clear() 해주기만 해도 테스트가 격리될 것으로 기대되었기 때문에, `@DirtiesContext`를 통해 불필요하게 Spring의 모든 컨텍스트를 초기화해주는 과정을 생략하고 싶었습니다!

threepebbles committed:
refactor: 테스트 패키지 구조를 프로덕션 패키지 구조와 동일하게 변경

threepebbles committed:
fix: 패키지명 올바르게 수정

threepebbles committed:
fix: SQL 쿼리문 오타 수정

- VALUE -> VALUES

threepebbles committed:
refactor: orElseThrow를 사용하여 가독성 개선

- 리뷰: https://github.com/woowacourse/spring-roomescape-admin/pull/321#discussion_r2059652789

threepebbles committed:
chore: 파일 끝에 개행 추가

- 리뷰: https://github.com/woowacourse/spring-roomescape-admin/pull/321#discussion_r2059654617

threepebbles committed:
refactor: SQL 쿼리문 내부 같은 필드에 대한 alias 이름 통일

- 리뷰: https://github.com/woowacourse/spring-roomescape-admin/pull/321#discussion_r2059665027

threepebbles committed:
refactor: SELECT 쿼리문에서 필요한 컬럼의 정보만 가져오도록 수정

threepebbles committed:
refactor: getString 대신 getTime 하도록 수정

- 리뷰: https://github.com/woowacourse/spring-roomescape-admin/pull/321#discussion_r2059666383

threepebbles committed:
refactor: RowMapper를 재사용 하도록 수정

- 리뷰: https://github.com/woowacourse/spring-roomescape-admin/pull/321#discussion_r2059667450

threepebbles committed:
test: test: 예약 API 테스트 중복 제거

- createReservationRequestJsonMap 메서드를 추출하여 중복된 요청 파라미터 생성 코드를 통합함
- 리뷰: https://github.com/woowacourse/spring-roomescape-admin/pull/321#discussion_r2059678314

threepebbles committed:
test: 중복된 예약 정보 삭제 기능 테스트를 하나로 통일

- 리뷰: https://github.com/woowacourse/spring-roomescape-admin/pull/321#discussion_r2059682087

threepebbles committed:
test: 예약 정보 저장 기능을 명확하게 테스트하도록 수정

- 조회 시 예외 발생 여부만 검증 -> 조회하여 저장된 데이터 값 모두 일치하는지 검증하도록 수정
- 리뷰: https://github.com/woowacourse/spring-roomescape-admin/pull/321#discussion_r2059690098

threepebbles committed:
test: 불필요한 Qualifier 설정 제거

- 리뷰: https://github.com/woowacourse/spring-roomescape-admin/pull/321#discussion_r2059694588

threepebbles committed:
remove: 코드 스타일 컨벤션 파일 제거

threepebbles committed:
refactor: 불변성을 위해 로컬 변수들에 final 추가

threepebbles commented:
안녕하세요 피케이! 헤일러입니다!

저번 리뷰 요청 때 [답변 드리지 못했던 내용](https://github.com/woowacourse/spring-roomescape-admin/pull/321#issuecomment-2831882383)과 함께 피드백 주신 내용 반영해보았습니다!

리뷰 잘 부탁드립니다 😄

threepebbles commented:
안녕하세요 피케이! 헤일러입니다!

저번 리뷰 요청 때 [답변 드리지 못했던 내용](https://github.com/woowacourse/spring-roomescape-admin/pull/321#issuecomment-2831882383)과 함께 피드백 주신 내용 반영해보았습니다!

리뷰 잘 부탁드립니다 😄 

(방금 실수로 커멘트 삭제를 눌러버려서 다시 입력했습니다 😅)

pkeugine commented on README.md:
```diff
@@ -1,23 +1,39 @@
-# Escape Room Back Office - 예약 관리 시스템
+# Escape Room System
```
이전 단계에서 작성해주신 답변 확인해보았습니다.
잘 남겨주셨네요 💯 💯 

**Q. 현재 상황에서 도메인 모델과 데이터베이스 엔티티를 분리하는 것이 큰 이점이 있을까요?**
헤일러가 남겨주신 '분리했을 때의 장/단점' 에서 저는 장점에 크게 공감하지 못하는 것 같습니다.
필드가 하나 줄어드는 것은 그리 크게 와닿는 장점이 아니네요.
하지만 단점은 저도 크다고 생각합니다. 변환해줘야 하고, 관리할 때 두 개 모두 관리해야 하죠.
따라서 불편함이 더 크기 때문에 분리의 이점이 딱히 없다고 생각합니다.

... 그런데 왜 `LocalDateTime` 리뷰를 남겼는가?
JDK 에서 제공하는 클래스 또는 api 에 최대한 익숙해지고, 어느 것을 활용하면 좋을지 고민하는 과정 자체가 의미있다고 생각했습니다. 헤일러의 말처럼 DB 의 컬럼 구조를 최대한 따라가는 방식 좋다고 생각합니다.

pkeugine commented on src/test/java/roomescape/reservation/domain/ReservationTest.java:
```diff
@@ -0,0 +1,49 @@
+    @Test
+    void 예약_시간이_null이면_예외가_발생한다() {
+        // given
+        final String name = null;
```
`name` 을 `null` 이 아닌 값으로 지정해야 더 목적에 맞는 테스트가 될 것 같네요.

threepebbles replied:
의미 있는 값으로 수정했습니다!

[8aa00e5](https://github.com/woowacourse/spring-roomescape-admin/pull/321/commits/8aa00e5b958228449d99bbbe1169410a24b8bafe)

pkeugine reviewed (changes requested):
안녕하세요 헤일러,
잘 수정해주셨고, 남긴 답변들도 잘 봤습니다.
몇 가지만 다시 확인해주세요.

threepebbles committed:
refactor: 불필요한 Qualifer 어노테이션 제거

threepebbles committed:
style: 프로젝트 파일 전체 reformat code 적용

threepebbles committed:
test: 테스트 목적에 맞는 given 값으로 수정

- name이 null이 아닌 의미있는 값으로 수정

threepebbles committed:
test: 불필요한 bean 이름 지정 제거

threepebbles committed:
test: 불변성을 위해 로컬 변수들에 final 추가

threepebbles committed:
test: 예약 정보 객체 생성 시 이름이 빈 문자열인 경우 예외 발생 테스트 추가

threepebbles commented:
안녕하세요 피케이! 헤일러입니다.
꼼꼼하게 피드백 해주신 덕분에 많이 배웠습니다! 감사합니다 👍 👍 

피드백 주신 내용 반영해보았습니다!
혹시 추가적으로 개선할 만한 사항이 있다면 피드백 주시면 감사드리겠습니다 😄

pkeugine reviewed (approved):
안녕하세요 헤일러,
리뷰 잘 반영해주셨네요.
이번 미션은 여기서 마무리하겠습니다.
다음 미션도 파이팅입니다 🔥
